### PR TITLE
Merge features/ioperation to master

### PR DESCRIPTION
**Pull request by mavasani**
Merge features/ioperation to master

### CODE REVIEW

**Comment by mavasani** [comment]
@jasonmalinowski @jinujoseph - as we discussed, I have rolled back the compiler toolset and analyzers used to 2.3.X with https://github.com/dotnet/roslyn/pull/22173/commits/96d2f97b8c815cb74ef550b76f886f406ce9f322. We can move to latest toolset and analyzers in dev15.5 after the code flows there.

File: NuGet.Config
Comment (jasonmalinowski) [inline NuGet.Config]: Redundant with line 28.

File: build/Rulesets/Roslyn_BuildRules.ruleset
Comment (jasonmalinowski) [inline build/Rulesets/Roslyn_BuildRules.ruleset]: Which one was this?

**Review by jasonmalinowski — APPROVED** [review]
Approving from the infrastructure perspective that this isn't bringing along some unexpected targets change which sometimes happens with big branch merges like this. I did not review any .cs or .vb content.

**Comment by jasonmalinowski** [comment]
@mavasani Looks like rolling the toolset back is causing some existing analyzer to catastrophically fail. 😦

**Comment by mavasani** [comment]
Yes Jason. https://github.com/dotnet/roslyn-analyzers/pull/1295 should fix this and we will publish use new analyzer package 2.3.0-beta2.

File: build/Rulesets/Roslyn_BuildRules.ruleset
Comment (mavasani) [reply build/Rulesets/Roslyn_BuildRules.ruleset]: Avoid empty allocations - uses an IOperation API that was removed, should be fixed when we move the analyzer to use new APIs.

**Comment by mavasani** [comment]
@dotnet/analyzer-ioperation @dotnet/roslyn-compiler The tests are now passing and this is ready to review.

**Comment by mavasani** [comment]
@dotnet/roslyn-compiler @dotnet/analyzer-ioperation Can I get reviews please?

**Comment by mavasani** [comment]
Merged after getting offline approval from leads.

**Pull request merged**
